src/state.ml

```diff
@@ -313,44 +313,46 @@ let pp ppf t =
 
 let compress s = s.session.compress
 
-(* TODO this depends on the cipher used.. *)
-let mtu config compress =
+let mtu config session =
   (* we assume to have a tun interface and the server send us a tun-mtu *)
   let tun_mtu =
     match Config.find Tun_mtu config with
     | None -> 1500 (* TODO "client_merge_server_config" should do this! *)
     | Some x -> x
-  in
-  let hmac_len =
-    (* TODO we now always have a Auth value in config -- revisit (also in light of #36) *)
-    Option.value ~default:`SHA1 (Config.find Auth config)
-    |> Mirage_crypto.Hash.digest_size
-  in
-  (* padding, done on replay_id + [timestamp] + compress + data *)
-  let static_key_mode = Config.mem Secret config in
-  let not_yet_padded_payload =
-    Packet.id_len (* sequence number *)
-    + (if static_key_mode then 4 else 0)
-    (* time stamp in static key mode *)
-    + if compress then 1 else 0
-  in
-  let hdrs =
-    2
-    (* hdr: 2 byte length *) + (if static_key_mode then 0 else 1)
-    (* 1 byte op + key *) + Packet.cipher_block_size
-    (* IV *) + hmac_len
-  in
-  (* now we know: tun_mtu - hdrs is space we have for data *)
-  let data = tun_mtu - hdrs in
-  (* data is pad ( not_yet_padded_payload + x ) - i.e. we're looking for the
-     closest bs-1 number, and subtract not_yet_padded_payload *)
-  let left = data mod Packet.cipher_block_size in
-  let r =
-    if left = pred Packet.cipher_block_size then data - not_yet_padded_payload
-    else data - succ left - not_yet_padded_payload
-  in
-  assert (r > 0);
-  r
+  and compress = session.compress in
+  match Config.get Cipher config with
+  | `AES_256_CBC ->
```

Comment:
Hmm I wonder how this interacts with tls-crypt which uses a fixed cipher and hmac algorithm :/

Reply:
I was wondering as well -- and guess we need both a data_mtu and a control_mtu -- since the former is used for any payload to be transferred via the wire; and the second needs to take the hmac etc. into account -- even if AEAD is used, the control channel (even for tls-auth) is using the auth. This means that the match on cipher is wrong.